docs/interfaces.md

```diff
@@ -0,0 +1,431 @@
+typedef struct internal_tlcc_state_t {
+        const char* INTEL_PUB_PEM, // Intel IAS verification key
+        
+        uint32_t sequence_number, // block number counter
+
+        // integrity metadata
+        kvs_t metadata_state,
+
+        // msp 
+        X509_STORE* root_certs_orderer, // root cert store for orderer org
+        X509_STORE* root_certs_apps, // root cert store for application orgs
+
+        // tlcc session
+        session_keys_t session_keys,
+};
+
+typedef struct kvs_t {};
+
+typedef struct session_keys_t {};
+```
+
+
+# ECC
+
+## Interface:
+The ECC interface is implemented through a "normal" chaincode interface. Methods with leading underscores
+are treated as FPC commands. Normal `invoke` invocations are forwarded to a FPC chaincode enclave.
+
+```go
+// chaincode interface (exposed to admin/clients) implemented by invoke
+func Invoke(stub shim.ChaincodeStubInterface) pb.Response {}
+
+// triggered by an admin
+CreateEnclave(params CC_Parameters) (Credentials, error)
+
+// key generatation 
+GenerateCCKeys(params CC_Parameters) (CCKey_registration_message, error)
+ExportCCKeys(credentals Credentials) (ExportMessage, error)
+ImportCCKeys(params CC_Parameters) (CCKey_registration_message, error)
+
+// chaincode invoke
+ChaincodeInvoke(transactionProposal []byte, ctx interface{}) (response, error)
+```
+
+```go
+// ECC must trigger this when sealedState is available using load() and ecc starts; this is not really a public interface
+createEnclaveWithSealedState(sealedState []byte) error
+
+// Storage interface
+// TODO how to store sealedState, where and under what identifier (e.g. chaincode_id);
+load(chaincode_id string) (sealedState []byte, err error)
+store(chaincode_id string, sealedState []byte) error
+```
+
+This interface is implemented by ECC to let a chaincode enclave call into the peer
+
+```c++
+
+// logging
+void ocall_print_string([in, string] const char *str);
+
+
+// state access
+// TBD does this key already contain the chaincode_id/namespace?
+void ocall_get_state(
+        [in, string] const char *key, 
+        [out, size=max_val_len] uint8_t *val, uint32_t max_val_len, [out] uint32_t *val_len,
+        [in, out] sgx_cmac_128bit_tag_t *cmac,
+        [user_check] void *u_shim_ctx);
+
+void ocall_put_state(
+        [in, string] const char *key,
+        [in, size=val_len] uint8_t *val, uint32_t val_len,
+        [user_check] void *u_shim_ctx);
+
+void ocall_get_state_by_partial_composite_key(
+        [in, string] const char *comp_key,
+        [out, size=max_len] uint8_t *values, uint32_t max_len, [out] uint32_t *values_len,
+        [in, out] sgx_cmac_128bit_tag_t *cmac,
+        [user_check] void *u_shim_ctx);
+
+
+// verify msp with the help of tlcc
+void ocall_verify_msp_id(
+        [in, string] const char *msp_id,
+        [in, size=32] uint8_t *nonce,
+        [out] bool isValid, 
+        [out] sgx_cmac_128bit_tag_t *cmac);
```

Comment:
This would be exactly the same EDL as TLCC has? (at least i see no reason to have two different .edls ...)  If so, maybe just reference here tlcc rather than redefining?

Reply:
TLCC would not need `ocall_get_state`? Ok, what is the best practise here? ECC needs the TLCC related to calls to perform TLCC `ecalls`. Having a single TLCC-integrity-metadata edl that contains, both, ecalls and ocalls, and is included by ECC and TLCC, would't the edg8er create stub code for all the functions in both enclaves? That is, we have used stub code in TLCC and ECC? Am I wrong here?

Reply:
oops, you are correct, the first three of course go to shim (and the fourth is gone)